### PR TITLE
Implement array based hash map `AHashMap`

### LINKED ISSUE #93663: AnimationTree causes performance drops even if no animation is running
### Tested versions

v4.3.beta2.mono.official [b75f0485b]

### System information

Godot v4.3.beta2.mono - Windows 10.0.19045 - Vulkan (Forward+) - dedicated NVIDIA GeForce RTX 3060 Ti (NVIDIA; 31.0.15.3623) - AMD Ryzen 7 3700X 8-Core Processor (16 Threads)

### Issue description

Whenever using Animation Tree (adding it to scene) even if no animation is running, it drops the FPS by ~50, so if you have 20, that is not much it will drop 1000 FPS.

image

Without Animation Tree
image

With Animation Tree (20 nodes)
image

### Steps to reproduce

I have a sample project that reproduces it run with the AnimationTrees and press F3 (Thanks @Calinou for the awesome plugin)

And after that delete the AnimationsTrees node and run it again and press F3.

image

### Minimal reproduction project (MRP)

[anim_tree.zip](https://github.com/user-attachments/files/16016153/anim_tree.zip)

### PR DESCRIPTION
**Pull request by Nazarwadim**
Currently, `HashMap` is implemented as a double linked list, which creates a significant problem with CPU cache pollution, as we store not only the necessary data but also pointers and allocator metadata. There are also costs for searching for free memory in the heap.

# Implementation view of list and array based hash map.
**ListHashMap:**

**AHashMap:**

# When to use HashMap instead of AHashMap:
- You need to keep an iterator or const pointer to Key and you intend to add/remove elements in the meantime.
- You need to preserve the insertion order when using erase.(Erase at the position of the deleted element inserts the element from the end)
```
      <-------------
    |               |
6 8 X 9 32 -1 5 -10 7 X X X 

6 8 7 9 32 -1 5 -10 X X X X
```
# What did I do in this pr:
I implemented the structure, added tests, and replaced maps that I thought needed to be replaced in the animation code.

# Benchmarks:
I made 3D animation benchmarks (many 3D models play the same animation):
[Animation_test.zip](https://github.com/godotengine/godot/files/15501000/Animation_test.zip)

ListHashMap(Master):
Project FPS: 77 (12.98 mspf)
Project FPS: 78 (12.82 mspf)
Project FPS: 78 (12.82 mspf)
Project FPS: 79 (12.65 mspf)

AHashMap(Current PR):
Project FPS: 103 (9.70 mspf)
Project FPS: 104 (9.61 mspf)
Project FPS: 103 (9.70 mspf)
Project FPS: 103 (9.70 mspf)
Project FPS: 102 (9.80 mspf)

Here you can see a 27% increase in performance. 

**List/array HashMaps benchmarks**
KeyValue<long, long>.
Benchmarking shows 2000% faster iteration, 25% faster random search, and 300% faster insertion compared to the current HashMap implementation. AHashMap<long, long> uses 2.5 times less memory than HashMap<long, long>.

**Projects in which I tested for regression** 
From demo projects Voxel, bullet shower, platformer2/3d, particles 2/3d, and some of my own projects.

BugSquad edit: Closes #93663

### CODE REVIEW

**Comment by dhoverml** [comment]
I did a brief look through the implementation, and there is quite a bit of duplicate code between this and the original HashMap implementation.

Additionally, the new implementation uses powers-of-2 in the modulo calculations, while the old implementation uses something called `fastmod`.  Since [integer division is slow](https://stackoverflow.com/a/27980441), and that almost every function uses modulo including in while/for loops, using powers-of-2 for modulo alone can improve performance quite a bit.  For all we know, that is where we are seeing the performance improvement, and not from storing the elements in an array.

I think applying the performance improvements (besides storing elements in an array) to HashMap should be done so that the _only_ difference is that AHashMap stores elements in an array.  This way we can confidently assert that the performance gain is from removing cache misses and not from something else.

**Comment by Nazarwadim** [comment]
@dhoverml see this PR : #90082

I just made benchmarks for animations for that PR. There is an improvement of 8-9% comparing with master. It should also be taken into account that there is a change in the hasher, which also has a significant improvement effect.

**Comment by dhoverml** [comment]
Oh nice, that's great!
Sorry if I missed it, but have you compared #90082 with this PR?

**Comment by Nazarwadim** [comment]
> Oh nice, that's great! Sorry if I missed it, but have you compared #90082 with this PR?

If this PR adds 23%, and that one adds 8%, then the difference will be 15%. Both they give about 25-26%.

**Review by Calinou — APPROVED** [review]
Tested locally (rebased on top of `master` https://github.com/godotengine/godot/commit/292e50e17e3f6e2509d3178a00204f964a907460), it works as expected. See https://github.com/godotengine/godot/pull/92838#pullrequestreview-2111592592 for animation benchmark results (including results when coupled with #92838).

File: core/templates/a_hash_map.h [hunk 1]
```diff
@@ -0,0 +1,663 @@
...
+				if (data.data != EMPTY_HASH) {
+					_insert_with_hash(data.hash, data.hash_to_key);
+				}
+			}
+		}
+
+		Memory::free_static(old_map_data);
+	}
+
+	_FORCE_INLINE_ int32_t _insert_element(const TKey &p_key, const TValue &p_value, uint32_t hash) {
```
Comment (AThousandShips) [inline core/templates/a_hash_map.h:209]:
```suggestion
	_FORCE_INLINE_ int32_t _insert_element(const TKey &p_key, const TValue &p_value, uint32_t p_hash) {
```

**Comment by Chaosus** [comment]
Looks interesting... and what the difference of this and `OAHashMap`? And where to use this and where `OAHashMap`? Can this be used in `AStar3D` class instead of `OAHashMap` for points to increase its performance?

**Comment by rossbridger** [comment]
I took a look at the hashmap implementation and haven't noticed any issue with the code (not a fan of robin hood hashing tho). I do have one suggestion, as @dhoverml pointed out, this `AHashMap` shares a lot of code with original `HashMap`, it is possible to unify the two classes by adding a new conditional template parameter, something like `template <typename TKey, typename TValue, typename Hasher = HashMapHasherDefault, typename Comparator = HashMapComparatorDefault<TKey>, `**bool IsArrayedHashMap=true**`> class HashMap`. It should be able to deduplicate the code.

**Comment by Nazarwadim** [comment]
@Chaosus 
> what the difference between this and OAHashMap?

The difference is that the elements in A are stored linearly as in a dynamic array. In OA, the elements are stored in the array at the position where their index matches the index where the hash is stored. 
In A is stored the index of the element in array next to the hash.

> And where to use this and where OAHashMap?

It is better to use OA where the size of the pair is < 16 bytes and the classes of keys and values ​​do not have constructors and destructors (they are trivial).

> Can it be used in AStar3D class instead of OAHashMap for points to increase its performance?

Yes, it can.

**Comment by Saul2022** [comment]
Testing on a s23 + qualcom adreno 740

Master  28-30 dropping even to 25 fps

This pr : 37-38 or 39 dropping to 36 , prettt good increase 😀, can't test the blend process ome because theres no android editor artifact. Using calinou test project

**Comment by rossbridger** [comment]
> @rossbridger This is not possible because the code is very different. Different iteration algorithm, different attributes, different implemented functions.

I think it's possible, you just need to use some C++11 template metaprogramming features like `std::conditional` to instantiate different data structures and algorithms at compile time as long as public function signatures are the same. That being said, if you don't think it would improve maintainability, it's fine having a separate class for .

Also, do you think we can replace the `HashMap` with `AHashMap` in [core/object/worker_thread_pool.h](https://github.com/godotengine/godot/blob/97b8ad1af0f2b4a216f6f1263bef4fbc69e56c7b/core/object/worker_thread_pool.h#L132) and [drivers/d3d12/rendering_device_driver_d3d12.h](https://github.com/godotengine/godot/blob/97b8ad1af0f2b4a216f6f1263bef4fbc69e56c7b/drivers/d3d12/rendering_device_driver_d3d12.h#L267) so that we can remove `Allocator` template parameter from `HashMap`? These are the only places with custom `Allocator` and I don't see the use of `HashMap` there requires preserving insertion order.

**Comment by Nazarwadim** [comment]
> Also, do you think we can replace the `HashMap` with `AHashMap` in [core/object/worker_thread_pool.h](https://github.com/godotengine/godot/blob/97b8ad1af0f2b4a216f6f1263bef4fbc69e56c7b/core/object/worker_thread_pool.h#L132) and [drivers/d3d12/rendering_device_driver_d3d12.h](https://github.com/godotengine/godot/blob/97b8ad1af0f2b4a216f6f1263bef4fbc69e56c7b/drivers/d3d12/rendering_device_driver_d3d12.h#L267)

I don't think it is necessary. Since `erase` is often used in ThreadPool (didn't check, but maybe it is faster in normal map), and in d3d12 `BarrierRequest` weighs a lot.

**Comment by Nazarwadim** [comment]
So, at this point I think it would be better to use this map only for animations and not for the whole engine to avoid regression and make the review process easier. And the use for all other cases should be done in other PRs after merging this PR.

**Comment by fire** [comment]
@TokageItLab for tracking animation performance improvements

**Comment by Nazarwadim** [comment]
@RandomShaper There is no limit to the size of an element. Just in my opinion, reallocating a large array can take more time than just creating a new element.

In the description, I wrote mandatory items. But I will try to add recommended ones.

**Comment by RandomShaper** [comment]
Any reasonable guideline like that, even if subject to specifics of each case, will help. (In any case, I still believe that every map (and every container class, for that matter) should have a bulleted list about its traits: iterator invalidation, sample usage situations (generic), etc., but that's aside.)

**Comment by Nazarwadim** [comment]
Made a description of the map in the header file.

**Comment by Saul2022** [comment]
Off topic , but i feel some curiosity why before it said 57 before and now in that same case says 76, without mentioning the after result because with them looks almost double the original performance of before( 99fps).. Is there any that improved that?

**Comment by Nazarwadim** [comment]
@Saul2022  This #97744 and this #92838 improved master performance, so I updated it. And #92838 actually improved performance a little better than the PR description says. 
Also, this week, I worked a on the use of array methods (`get_index`, `get_by_index`) in the animation code, so as not to use the map lookup in bottlenecks.

**Review by RandomShaper — APPROVED** [review]

File: core/templates/a_hash_map.h [hunk 2]
```diff
@@ -0,0 +1,732 @@
...
+		HashMapData c_data;
+		c_data.hash = p_hash;
+		c_data.hash_to_key = p_index;
+
+		while (true) {
+			if (map_data[pos].data == EMPTY_HASH) {
+#ifdef DEV_ENABLED
+				if (unlikely(distance > 12)) {
+					WARN_PRINT("Excessive collision count (" +
+							itos(distance) + "), is the right hash function being used?\nClass Name:" + __PRETTY_FUNCTION__);
```
Comment (Repiteo) [inline core/templates/a_hash_map.h:190]: `__PRETTY_FUNCTION__` isn't guaranteed to be included in an MSVC context, nor is it used anywhere else in the project. It shouldn't be necessary either, as `__FUNCTION__` is always passed implicitly in error macros.
```suggestion
							itos(distance) + "), is the right hash function being used?);
```

**Review by Repiteo — CHANGES_REQUESTED** [review]

File: core/templates/a_hash_map.h (hunk 2 above)
Comment (Nazarwadim) [reply core/templates/a_hash_map.h:190]: I added it because it is necessary to be able to understand in which template there is a problem.

Comment (Nazarwadim) [reply core/templates/a_hash_map.h:190]: Although you might be right because it can be easily viewed with a debugger.

Comment (Nazarwadim) [reply core/templates/a_hash_map.h]: Done

**Review by Repiteo — APPROVED** [review]

**Comment by Repiteo** [comment]
Thanks!

**Pull request merged**